### PR TITLE
use containers we expect to start for wait condition

### PR DESCRIPTION
**Pull request by ndeloof**
as we check dependency service is running/healthy, use containers we created for the service so that we can detect failure to start, and not wait with no end for a dead service to become healthy.

**Related issue**
closes https://github.com/docker/compose/issues/10200

**(not mandatory) A picture of a cute animal, if possible in relation to what you did**
image

### CODE REVIEW

**Comment by rborn-tx** [comment]
Hello @ndeloof. I've run some tests on your branch and I've found some possible issues that I wanted to let you know. If I should instead report this as part of the original issue ticket, please let me know, okay?

### Test 1

```yaml
version: "2.4"
services:
  # Long-lived container.
  hello1:
    image: "alpine"
    command: ["sh", "-c", "echo 'Long-lived container, no health check'; sleep 1h;"]

  # Short-lived container.
  hello2:
    image: "alpine"
    command: ["echo", "Short-lived container, no health check"]
```

Execution:

```
$ compose --file docker-compose-mixed-nohc.yml up --detach --remove-orphans --wait
[+] Running 2/3
 ⠿ Network aktualizr-tests_default     Created                                                    0.0s
 ⠿ Container aktualizr-tests-hello2-1  Waiting                                                   15.8s
 ⠿ Container aktualizr-tests-hello1-1  Started                                                    0.4s
Killed
```
Hitting ^C did not stop docker-compose which had to be killed.


### Test 2

```yaml
version: "2.4"
services:
  # Short-lived container.
  hello:
    image: "alpine"
    command: ["echo", "Short-lived container, no health check"]
```

Execution:

```
$ compose --file docker-compose-shortlife-nohc.yml up --detach --remove-orphans --wait
[+] Running 1/2
 ⠿ Network aktualizr-tests_default    Created                                                    0.0s
 ⠿ Container aktualizr-tests-hello-1  Waiting                                                    0.8s
container aktualizr-tests-hello-1 exited (0)

$ echo $?
1
```

Notice in this case the `--wait` didn't cause docker-compose to wait forever, but the exit code of the program was non-zero indicating something went wrong.

### Test 3

```yaml
# Run multiple short-lived containers.
version: "2.4"
services:
  hello1:
    image: "alpine"
    command: ["echo", "Short-lived container (1), no health check"]
  hello2:
    image: "alpine"
    command: ["echo", "Short-lived container (2), no health check"]
  hello3:
    image: "alpine"
    command: ["echo", "Short-lived container (3), no health check"]
```

Execution:

```
$ compose --file docker-compose-shortlife-nohc3.yml up --detach --remove-orphans --wait
[+] Running 3/4
 ⠿ Network aktualizr-tests_default     Created                                                          0.0s
 ⠿ Container aktualizr-tests-hello1-1  Waiting                                                         26.9s
 ⠿ Container aktualizr-tests-hello2-1  Started                                                          0.5s
 ⠿ Container aktualizr-tests-hello3-1  Started                                                          0.6s
Killed
```

Like with Test 1, hitting ^C did not stop docker-compose which had to be killed.

**Comment by ndeloof** [comment]
@rborn-tx thanks, nice feedback. I'll try to convert those examples into e2e tests so we have better coverage for this feature

**Comment by ndeloof** [comment]
I fixed reported corner cases.
For second one ("Test 2") service command immediately complete, `--wait` detect container exited and report failure to become "running". This sounds a reasonable behavior as service is expected to run until interrupted, not to complete this way (otherwise, --wait makes no sense)

File: pkg/compose/convergence.go [hunk 1]
```diff
@@ -291,10 +290,7 @@ func (s *composeService) waitDependencies(ctx context.Context, project *types.Pr
 			continue
 		}
 
-		containers, err := s.getContainers(ctx, project.Name, oneOffExclude, false, dep)
-		if err != nil {
-			return err
-		}
+		containers = containers.filter(isService(dep))
```
Comment (glours) [inline pkg/compose/convergence.go:293]: Shouldn't we use a dedicated variable to manage filtered containers? I wondering if we won't loose potential dependencies by re-assigning `containers`

**Review by glours — CHANGES_REQUESTED** [review]
Not sure, but we may have a potential side effect in the dependencies check loop

File: pkg/compose/convergence.go (hunk 1 above)
Comment (ndeloof) [reply pkg/compose/convergence.go:293]: indeed, nice catch - I've been lazy here

**Comment by kaos14** [comment]
> I fixed reported corner cases. For second one ("Test 2") service command immediately complete, `--wait` detect container exited and report failure to become "running". This sounds a reasonable behavior as service is expected to run until interrupted, not to complete this way (otherwise, --wait makes no sense)

I disagree. I have multiple services in my docker-compose.yml, that I do want compose to wait for. At the same time I have a service that does its job and exits with 0 code. This change broke my setup and I had to add artificial sleep at the end of the script. I think `--wait` should take into account the `restart` option and not error out in Test 2.

**Comment by kaos14** [comment]
@ndeloof What did you mean "I fixed reported corner cases."? None of them works in 2.16. Is this as expected?
